### PR TITLE
Python: New command execution sinks

### PR DESCRIPTION
**Pull request by am0o0**
What is new?
JsonPickle library Code execution sinks
Pytorch library Code execution sinks
Pexpect library Command Execution and Secondary server cmd injection
AsyncSsh library Secondary server cmd injection
Netmiko library Secondary server cmd injection
Scrapli library Secondary server cmd injection
Twisted library Secondary server cmd injection
Ssh2-python library Secondary server cmd injection
pandas library DataFrame Code execution sinks

What has changed?
Upgrade paramiko query to Secondary server command execution query which attackers can execute commands on other than the primary server. it is in the experimental directory.
for the paramiko query, it has added `proxyCommand` as a `SystemCommandExecution` because it executes commands on the primary server.
Upgraded `Fabric` framework and added proxy_command as a `SystemCommandExecution`, I didn't change the sinks of this framework to Secondary server command execution because it is not in an experimental library, otherwise the `run` and `sudo` functions are `SecondaryCommandInjection` and `local` function is `SystemCommandExecution`. I only simplified the framework structure with new higher-level APIs and added `SystemCommandExecution`  new sinks.

Also, I tried my best to use inline tests everywhere so you can review this PR more easily. :)

### CODE REVIEW

**Comment by ghsecuritylab** [comment]
Hello am0o0 👋
You have submitted this pull request as a bug bounty report in the [github/securitylab](https://github.com/github/securitylab) repository and therefore this pull request has been put into draft state to give time for the GitHub Security Lab to assess the PR. When GitHub Security Lab has finished assessing your pull request, it will be marked automatically as `Ready for review`. Until then, please don't change the draft state.

In the meantime, feel free to make changes to the pull request. If you'd like to maximize payout for your this and future submissions, here are a few general guidelines, that we might take into consideration when reviewing a submission.

- the submission models widely-used frameworks/libraries
- the vulnerability modeled in the submission is impactful
- the submission finds new true positive vulnerabilities
- the submission finds very few false positives
- code in the submission is easy to read and will be easy to maintain
- documentation is written clearly, highlighting the impact of the issue it finds and is written without grammatical or other errors. The code samples clearly show the vulnerability
- the submission includes tests, change note etc.

Please note that these are guidelines, not rules. Since we have a lot of different types of submissions, the guidelines might vary for each submission.

Happy hacking!

**Comment by tausbn** [comment]
Tests were failing to compile, so I rebased this on top of current `main`. I'm in the process of reviewing it now, but the fact that it targets our main libraries (and not `experimental`) means a greater level of scrutiny is required, and it'll therefore take a bit longer than expected.

**Comment by tausbn** [comment]
Urgh. Rebasing may have been a mistake (I ended up a co-author on all commits, which I wasn't expecting). Feel free to force push your original commits instead, and then merging in `main`.

**Comment by tausbn** [comment]
> @tausbn I think I should add some predicates for classes that extend the `SystemCommandExecution::Range` like the `SubprocessPopenCall` [class](https://github.com/github/codeql/blob/85e71c30dc97dd3771a21b8d63243fd9691c1b48/python/ql/lib/semmle/python/frameworks/Stdlib.qll#L1195). do you think I should do it now? or should I wait for the second review round?

Feel free to add them now, if you like. 👍 

> @tausbn Also please let me know if you prefer I move most of the new non-experimental codeql library files under the experimental directory.

No, I think it's fine the way it is. From what I've seen so far, your modelling is good, and only a few changes here and there should be needed to make it ready to merge. 🙂 

One thing I do need to mention is this change: https://github.com/github/codeql/pull/15715/files#diff-950dae083553f4d1115143425b3e4816da96a333a4751463eda140c20156ae5cL97

The predicate in question is used elsewhere, and so its removal is causing the tests to fail (as we can't even compile all the queries). Specifically, It's this file that uses it: https://github.com/github/codeql/blob/main/python/ql/src/meta/ClassHierarchy/Find.ql#L327

File: python/ql/src/meta/ClassHierarchy/Find.ql [hunk 1]
```diff
@@ -328,7 +328,7 @@ class FabricConnection extends FindSubclassesSpec {
   FabricConnection() { this = "fabric.connection.Connection~Subclass" }
 
   override API::Node getAlreadyModeledClass() {
-    result = FabricV2::Fabric::Connection::ConnectionClass::classRef()
+    result  = any(FabricV2::Fabric::Connection::ConnectionClass::Instance i)
```
Comment (tausbn) [inline python/ql/src/meta/ClassHierarchy/Find.ql]: This is not right. `classRef` should be a reference to the _class_, whereas `Instance` is for _instances_ of said class. These are not the same thing, and this is important for this file, as it is explicitly about finding references to (subclasses of) known classes, _not_ instances of those classes.

Please just put `classRef()` back in place in `Fabric.qll`.

(Also, don't forget to autoformat your code. This file and `Torch.qll` failed the autoformat check.)

File: python/ql/src/experimental/semmle/python/security/SecondaryServerCmdInjection.qll [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import codeql.util.Unit
+import experimental.semmle.python.Concepts
+
+module SecondaryCommandInjectionConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  predicate isSink(DataFlow::Node sink) { sink instanceof SecondaryCommandInjection }
+}
+
+/** Global taint-tracking for detecting "paramiko command injection" vulnerabilities. */
```
Comment (tausbn) [inline python/ql/src/experimental/semmle/python/security/SecondaryServerCmdInjection.qll:15]: More than just paramiko, no?

**Review by tausbn — CHANGES_REQUESTED** [review]
Thank you for your submission. Before we can merge this, there are a few comments that need to be addressed. 

Once you've addressed the requested changes, I'll re-run the tests and also kick off a performance evaluation (which is needed because you've made changes to files outside the `experimental` subdirectory).

Finally, this would probably have been better as two separate PRs, as the command execution sinks are mostly disjoint from your work on secondary command injection.

**Review by tausbn — DISMISSED** [review]
Thank you for the changes. I think this looks good. 👍 
However, there's now a merge conflict on the `paramiko.expected` file (which was deleted).

Could you kindly resolve that conflict? Then I can run the final performance tests, and (assuming there are no issues with performance) merge the PR.

**Comment by tausbn** [comment]
I think you solved the merge conflict (at least the tests can run now), but I'm not too sure about the revert in the latest commit.

Also:
```
python/ql/src/experimental/semmle/python/security/RemoteCommandExecution.qll would change by autoformatting.
```
(So many hoops to jump through...)

**Review by tausbn — APPROVED** [review]
Performance looks good! In it goes. 👍

**Pull request merged**